### PR TITLE
feat: add code coverage report support

### PR DESCRIPTION
**Pull request by sguiheux**
1. Description
1. Related issues
1. About tests
1. Mentions

@ovh/cds

### CODE REVIEW

File: engine/api/workflow_queue_test.go [hunk 1]
```diff
@@ -17,18 +17,24 @@ import (
 	"github.com/ovh/venom"
 	"github.com/stretchr/testify/assert"
 
+	"bytes"
```
Comment (golangci) [inline engine/api/workflow_queue_test.go]: File is not goimports-ed

File: engine/api/workflow_queue_test.go [hunk 2]
```diff
@@ -716,11 +722,273 @@ func Test_postWorkflowJobArtifactHandler(t *testing.T) {
...
+
+	// Create a workflow run
+	wrToTest, _, errT := workflow.ManualRun(context.Background(), db, api.Cache, p, &w, &sdk.WorkflowNodeRunManual{
+		Payload: map[string]string{
+			"git.branch": "my-branch",
+		},
+	}, nil)
+	assert.NoError(t, errT)
+
+	wrr, err := workflow.LoadRunByID(db, wrToTest.ID, workflow.LoadRunOptions{})
```
Comment (golangci) [inline engine/api/workflow_queue_test.go]: ineffectual assignment to `err`

File: engine/api/workflow_queue_test.go [hunk 3]
```diff
@@ -716,11 +722,273 @@ func Test_postWorkflowJobArtifactHandler(t *testing.T) {
...
+
+	// Create user
+	u, pass := assets.InsertAdminUser(api.mustDB())
+
+	// Create project
+	key := sdk.RandomString(10)
+	proj := assets.InsertTestProject(t, db, api.Cache, key, key, u)
+
+	// add group
+	group.InsertUserInGroup(api.mustDB(), proj.ProjectGroups[0].Group.ID, u.ID, true)
```
Comment (golangci) [inline engine/api/workflow_queue_test.go]: Error return value of `group.InsertUserInGroup` is not checked

File: engine/api/workflow_queue_test.go [hunk 4]
```diff
@@ -716,11 +722,273 @@ func Test_postWorkflowJobArtifactHandler(t *testing.T) {
...
+	}
+	p, err := project.Load(db, api.Cache, proj.Key, u, project.LoadOptions.WithPipelines, project.LoadOptions.WithApplications)
+	assert.NoError(t, err)
+	assert.NoError(t, workflow.Insert(db, api.Cache, &w, p, u))
+
+	repositoryService := services.NewRepository(func() *gorp.DbMap {
+		return db
+	}, api.Cache)
+	mockVCSservice := &sdk.Service{Name: "TestInsertNewCodeCoverageReport", Type: services.TypeVCS}
+	repositoryService.Delete(mockVCSservice)
```
Comment (golangci) [inline engine/api/workflow_queue_test.go]: Error return value of `repositoryService.Delete` is not checked

**Comment by ovh-cds** [comment]
CDS Report ut-engine#5053.0 ✘  
* Stage 1
  * Engine Test with DB ✘  
Unit Tests Report
* github.com/ovh/cds/cli


* github.com/ovh/cds/engine/api















































































  * Test_putPipelineImportHandler ✘

























































































  * Test_putWorkflowImportHandler ✘














  * Test_getWorkflowNodeRunHandler ✘

















* github.com/ovh/cds/engine/api/application








* github.com/ovh/cds/engine/api/environment




* github.com/ovh/cds/engine/api/keys



* github.com/ovh/cds/engine/api/migrate


* github.com/ovh/cds/engine/api/pipeline






* github.com/ovh/cds/engine/api/plugin


* github.com/ovh/cds/engine/api/poller




* github.com/ovh/cds/engine/api/project









* github.com/ovh/cds/engine/api/scheduler












* github.com/ovh/cds/engine/api/secret






* github.com/ovh/cds/engine/api/user








* github.com/ovh/cds/engine/api/warning















* github.com/ovh/cds/engine/api/worker










* github.com/ovh/cds/engine/api/workflow












































* github.com/ovh/cds/engine/hatchery/swarm



















* github.com/ovh/cds/engine/hooks






* github.com/ovh/cds/engine/repositories


* github.com/ovh/cds/engine/vcs











* github.com/ovh/cds/engine/vcs/bitbucket


















* github.com/ovh/cds/engine/vcs/github













* github.com/ovh/cds/engine/vcs/gitlab









* github.com/ovh/cds/engine/worker






* github.com/ovh/cds/sdk















* github.com/ovh/cds/sdk/exportentities




























* github.com/ovh/cds/sdk/interpolate

























* github.com/ovh/cds/sdk/luascript




* github.com/ovh/cds/sdk/plugin/dummy/caller


* github.com/ovh/cds/sdk/vcs/git

**Comment by ovh-cds** [comment]
CDS Report build-ui#5053.0 ✘  
* Compile
  * Build UI ✘  
  * TU model ✔ 
  * TU service ✔ 
  * TU shared ✔ 
  * TU views ✔

File: engine/api/workflow/dao_node_run.go [hunk 5]
```diff
@@ -75,6 +77,51 @@ func LoadNodeRun(db gorp.SqlExecutor, projectkey, workflowname string, number, i
...
+	join workflow on workflow.id = workflow_run.workflow_id
+	where workflow_node_run_job.id = $1`, nodeRunFields, testsField)
+
+	if err := db.SelectOne(&rr, query, nodeJobRunID); err != nil {
+		return nil, sdk.WrapError(err, "workflow.LoadNodeRunByNodeJobID> Unable to load workflow_node_run node_job_id=%d", nodeJobRunID)
+	}
+
+	r, err := fromDBNodeRun(rr, loadOpts)
+	if err != nil {
+		return nil, sdk.WrapError(err, "LoadNodeRun>")
```
Comment (bnjjj) [inline engine/api/workflow/dao_node_run.go]: s/LoadNodeRun/LoadNodeRunByNodeJobID

File: ui/src/app/views/workflow/run/node/test/tests.result.html [hunk 6]
```diff
@@ -1,26 +1,75 @@
...
-            <div class="right aligned five wide column">
-                <a class="pointing" (click)="filter='all'">
-                    {{tests.total}}
-                    <span *ngIf="tests.total > 1">{{'pipeline_run_test_total_title_s' | translate}}</span>
-                    <span *ngIf="tests.total < 2">{{'pipeline_run_test_total_title' | translate}}</span>
-                </a>
+            <div class="ui row" *ngIf="coverage">
+                <div class="right aligned five wide column">
+                    <span *ngIf="coverage.report.total_lines > 0" suiPopup popupText="{{coverage.report.covered_lines}}/{{coverage.report.total_lines}}">
+                        Covered lines: {{percentLines}}%
```
Comment (bnjjj) [inline ui/src/app/views/workflow/run/node/test/tests.result.html]: traduction

File: ui/src/app/views/workflow/run/node/test/tests.result.html [hunk 7]
```diff
@@ -1,26 +1,75 @@
...
+                        <span *ngIf="!currentBranchTrendLines && defaultBranchTrendLines" suiPopup popupText="{{'coverage_trend_default_branch' | translate }}">
+                            <i class="arrow icon" [class.down]="defaultBranchTrendLines > percentLines" [class.red]="defaultBranchTrendLines > percentLines"
+                               [class.up]="defaultBranchTrendLines < percentLines" [class.green]="defaultBranchTrendLines < percentLines"
+                            ></i>
+                        </span>
+                    </span>
+                </div>
+                <div class="centered five wide column">
+                    <span *ngIf="coverage.report.total_functions > 0" suiPopup popupText="{{coverage.report.covered_functions}}/{{coverage.report.total_functions}}">
+                        Covered functions: {{percentFunctions}}%
```
Comment (bnjjj) [inline ui/src/app/views/workflow/run/node/test/tests.result.html]: trad

File: ui/src/app/views/workflow/run/node/test/tests.result.html [hunk 8]
```diff
@@ -1,26 +1,75 @@
...
+                        <span *ngIf="currentBranchTrendFunctions" suiPopup popupText="{{'coverage_trend_current_branch' | translate }}">
+                            <i class="arrow icon" [class.down]="currentBranchTrendFunctions > percentFunctions" [class.red]="currentBranchTrendFunctions > percentFunctions"
```
Comment (bnjjj) [inline ui/src/app/views/workflow/run/node/test/tests.result.html]: add equals icon when no changes

File: ui/src/app/views/workflow/run/node/test/tests.result.html [hunk 9]
```diff
@@ -1,26 +1,75 @@
...
+                        <span *ngIf="!currentBranchTrendFunctions && defaultBranchTrendFunctions" suiPopup popupText="{{'coverage_trend_default_branch' | translate }}">
+                            <i class="arrow icon" [class.down]="defaultBranchTrendLines > percentFunctions" [class.red]="defaultBranchTrendLines > percentFunctions"
+                               [class.up]="defaultBranchTrendLines < percentFunctions" [class.green]="defaultBranchTrendLines < percentFunctions"
+                            ></i>
+                        </span>
+                    </span>
+                </div>
+                <div class="five wide column">
+                    <span *ngIf="coverage.report.total_branches > 0" suiPopup popupText="{{coverage.report.covered_branches}}/{{coverage.report.total_branches}}">
+                        Covered branches: {{percentBranches}}%
```
Comment (bnjjj) [inline ui/src/app/views/workflow/run/node/test/tests.result.html]: trad

**Comment by ovh-cds** [comment]
CDS Report ut-engine#5058.0 ✘  
* Stage 1
  * Engine Test with DB ✘  
Unit Tests Report
* github.com/ovh/cds/cli


* github.com/ovh/cds/engine/api
























































































































































































  * Test_getWorkflowNodeRunHandler ✘

















* github.com/ovh/cds/engine/api/application








* github.com/ovh/cds/engine/api/environment




* github.com/ovh/cds/engine/api/keys



* github.com/ovh/cds/engine/api/migrate


* github.com/ovh/cds/engine/api/pipeline






* github.com/ovh/cds/engine/api/plugin


* github.com/ovh/cds/engine/api/poller




* github.com/ovh/cds/engine/api/project









* github.com/ovh/cds/engine/api/scheduler












* github.com/ovh/cds/engine/api/secret






* github.com/ovh/cds/engine/api/user








* github.com/ovh/cds/engine/api/warning















* github.com/ovh/cds/engine/api/worker










* github.com/ovh/cds/engine/api/workflow












































* github.com/ovh/cds/engine/hatchery/swarm



















* github.com/ovh/cds/engine/hooks






* github.com/ovh/cds/engine/repositories


* github.com/ovh/cds/engine/vcs











* github.com/ovh/cds/engine/vcs/bitbucket


















* github.com/ovh/cds/engine/vcs/github













* github.com/ovh/cds/engine/vcs/gitlab









* github.com/ovh/cds/engine/worker






* github.com/ovh/cds/sdk















* github.com/ovh/cds/sdk/exportentities




























* github.com/ovh/cds/sdk/interpolate

























* github.com/ovh/cds/sdk/luascript




* github.com/ovh/cds/sdk/plugin/dummy/caller


* github.com/ovh/cds/sdk/vcs/git

File: engine/api/workflow/dao_coverage.go [hunk 10]
```diff
@@ -0,0 +1,197 @@
...
+	"github.com/ovh/cds/engine/api/cache"
+	"github.com/ovh/cds/engine/api/database/gorpmapping"
+	"github.com/ovh/cds/engine/api/repositoriesmanager"
+	"github.com/ovh/cds/sdk"
+)
+
+func loadPreviousCoverageReport(db gorp.SqlExecutor, workflowID int64, runNumber int64, repository string, branch string) (sdk.WorkflowNodeRunCoverage, error) {
+	query := `
+    SELECT * FROM (
+      SELECT * from workflow_node_run_coverage
```
Comment (fsamin) [inline engine/api/workflow/dao_coverage.go]: you can limit 1 just after the order without the sub request

File: engine/api/workflow/dao_coverage.go [hunk 11]
```diff
@@ -0,0 +1,197 @@
...
+		return sdk.WorkflowNodeRunCoverage{}, sdk.WrapError(err, "LoadPreviousCoverageReport> Unable to load previous coverage")
+	}
+
+	return sdk.WorkflowNodeRunCoverage(cov), nil
+}
+
+func loadLatestCoverageReport(db gorp.SqlExecutor, workflowID int64, repository string, branch string) (sdk.WorkflowNodeRunCoverage, error) {
+	query := `
+    SELECT * FROM (
+      SELECT * from workflow_node_run_coverage
```
Comment (fsamin) [inline engine/api/workflow/dao_coverage.go]: idem

File: engine/api/workflow/dao_coverage.go [hunk 12]
```diff
@@ -0,0 +1,197 @@
...
+	"github.com/sguiheux/go-coverage"
+
+	"github.com/ovh/cds/engine/api/cache"
+	"github.com/ovh/cds/engine/api/database/gorpmapping"
+	"github.com/ovh/cds/engine/api/repositoriesmanager"
+	"github.com/ovh/cds/sdk"
+)
+
+func loadPreviousCoverageReport(db gorp.SqlExecutor, workflowID int64, runNumber int64, repository string, branch string) (sdk.WorkflowNodeRunCoverage, error) {
+	query := `
```
Comment (fsamin) [inline engine/api/workflow/dao_coverage.go]: we should be able to handle coverage per application

**Pull request merged**